### ISSUE TITLE
Starting dockerized OneProvider fails if backend storage does not allow file modifications

### ISSUE THREAD

**Issue by jstarek** [onedata/onedata#10]
When trying to start OneProvider from the dockerized scenario 1.0, it will fail to start when using a file system backend that does not allow file modifications.

We observed this problem with OneData RC14, using an NFS-mounted dCache 2.16 as a backend. Upon startup, a 100 byte test file containing random ASCII characters was successfully written to the storage space: 
```# ls -al /data/storage/
total 2
drwxr-xr-x 3 root root 512 May  4 15:41 .
drwxr-xr-x 5 root root 512 May  4 15:40 ..
-rw-rw-rw- 1 root root 100 May  4 15:41 cemumsktjrpnxnkegizpitnxzdvtdorr
```

But subsequently, the startup scripts seem to try to modify that file, which fails.  Console output is:
```3.0.0-rc14: Pulling from onedata/oneprovider
Digest: sha256:65dbf6b494719c51e05099e0c40c7da8724ae9e4f09a168b92f4187cdd2a3589
Status: Image is up to date for onedata/oneprovider:3.0.0-rc14
Creating oneprovider-1
Attaching to oneprovider-1
oneprovider-1                | Starting op_panel	[  OK  ]
oneprovider-1                | 
oneprovider-1                | Configuring oneprovider:
oneprovider-1                | * service_couchbase: configure
oneprovider-1                | * service_couchbase: start
oneprovider-1                | * service_couchbase: wait_for_init
oneprovider-1                | * service_couchbase: init_cluster
oneprovider-1                | * service_couchbase: rebalance_cluster
oneprovider-1                | * service_couchbase: status
oneprovider-1                | * service: save
oneprovider-1                | * service_cluster_manager: configure
oneprovider-1                | * service_cluster_manager: stop
oneprovider-1                | * service_cluster_manager: start
oneprovider-1                | * service_cluster_manager: status
oneprovider-1                | * service_op_worker: configure
oneprovider-1                | * service_op_worker: setup_certs
oneprovider-1                | * service_op_worker: stop
oneprovider-1                | * service_op_worker: start
oneprovider-1                | * service_op_worker: wait_for_init
oneprovider-1                | * service_op_worker: status
oneprovider-1                | * service_op_worker: add_storages
oneprovider-1                | 
oneprovider-1                | Error: Service Error
oneprovider-1                | Description: Action 'deploy' for a service 'oneprovider' terminated with an error.
oneprovider-1                | Module: service_op_worker
oneprovider-1                | Function: add_storages
oneprovider-1                | Hosts: dcache-onedata-01.desy.de
oneprovider-1                | For more information please check the logs.
oneprovider-1 exited with code 1
```

Attached is the [op_panel_error.log.txt](https://github.com/onedata/onedata/files/976281/op_panel_error.log.txt) with more details about the problem. We can provide a tcpdump capture of the relevant NFS traffic if needed.

In general, dCache treats files as immutable once they are CLOSEd. The expectation is that client software that needs to modify a file does so by replacing the existing file, like e.g. the vim editor does. If mutability of files is tested during system startup, more detailed information (is this a required property of any file system backend?) and different error handling might be beneficial.

**Comment by groundnuty** [comment]
Thank you for a detailed description of the problem. Those are quirks that pop up now and then with very specific scenarios, like this one. We will need to discuss the support for *immutable* file systems and will get back to you. Up till now we just distinguished between rw/ro storages.

**Comment by groundnuty** [comment]
@jstarek 
Am I right to assume that they deployment you are aiming at, is a single node provider installation? (not multiple oneprovider nodes)

**Comment by jstarek** [comment]
Yes, for the moment we are looking at a single node installation. But in general, a multinode installation should definitely be possible as well.

**Comment by groundnuty** [comment]
@jstarek 
we managed to pinpoint the problem. as for now actually, we do not provider any official support for immutable filesystems. When deploying a multi-node provider the storage detection algorithm first creates a file and then tries to modify it from every other node. Hence in this scenario, it will fail when dealing with immutable filesystems.
The fact that it was working in a single node scenario was just by accident since the single node behavior of the algorithm was not consistent with the multi-node situation. A well-meaning developer fixed that lack of consistency in rc14, thus resulting in this bug issue.

We can reverse this change, although that still leaves the problem of multi-node provider working with dCache.

**Comment by groundnuty** [comment]
@jstarek 
We released a public nightly build that addresses this issue:
`docker image: onedata/oneprovider:nightly-VFS-3213`
Please confirm if that fix resolves the problem on your end. If so, it will be released as part of rc15, that would focus mainly on similar fixes. This build is based on our development branch, it can slightly differ from what's described in documentation of rc14.

**Comment by paulmillar** [comment]
@groundnuty, hijacking the ticket somewhat: you mention "first creates a file and then tries to modify it from every other node." Is this intended to support writing the file's data to multiple nodes in parallel?  If so, these writes would be concurrent and non-overlapping, if I've understood correctly.

**Comment by jstarek** [comment]
@groundnuty , thanks for providing access to the nightly build. It starts up successfully on our dCache test system.

In order to properly test if everything works in this combination, could you please give some examples for situations in which the Oneprovider actually needs a mutable file system? Thanks a lot!

**Comment by groundnuty** [comment]
@jstarek with this patch, the mutable filesystem should not need it at all (at least for storage detection process for sure) There ae just 2 cases to consider: 1) single-node oneprovider installation, 2) multi-node. The second case never worked until this patch.

**Comment by jstarek** [comment]
So, I've only now gotten around to installing and testing RC15. While it starts up nicely now, I observe file uploads failing if files are larger than 1 MB. I suppose that the web interface does some sort of chunked uploading? Tracing network activity during such an upload (see attached tcpdump file) shows that OneProvider does CLOSE calls after touching (creating) the file, and subsequently, there are sequences of GETATTR, OPEN and CLOSE calls during the transfer. This of course fails on backends with immutable FSs. So, even for a single-node instance, it looks like mutability is required...? Can the transfer mode be configured in any way?

**Comment by luman75** [comment]
The issue should be corrected as of release 17.06.0-beta4, we would appreciate if we you would try to deploy the newer version and check if it works on your write once storage.

**Comment by paulmillar** [comment]
Thanks for the feedback!

Jürgen's currently away, but should be back next week.  Hopefully we'll be able to confirm the new version fixes the problem then.

**Comment by jstarek** [comment]
Hi all, thanks for the work on this issue! I am seeing problems with starting the new RC, which I've detailed in our test mailing list under Lukasz's reply.

**Comment by jstarek** [comment]
@luman75 , could you please alert us when there are news with regard to the startup problem in 17.06.0-beta4? We're eager to try out the new version.

**Comment by groundnuty** [comment]
@jstarek as the newest Oneprovider releases include the possibility to mark storage as read-only. Could you please test this against the newest release `17.06.0-beta6`?

**Comment by jstarek** [comment]
Thanks for the new beta. I just tested it, and unfortunately, there are still some issues.

In principle, the container starts up (although it takes far longer than before, several minutes compared to a few seconds). After the message "Congratulations! oneprovider has been successfully started.", I get some stack traces and error messages on stdout. In order to help you debug those, I attached the relevant logs to this comment.

I will be away from office for a while, so I'll ask my colleagues to take over this issue while I'm away.

[crashreportlogs.zip](https://github.com/onedata/onedata/files/1169506/crashreportlogs.zip)

**Comment by paulmillar** [comment]
Some further information: with tcpdump, I see the following suspicious behaviour:

```
12:36:27.835191 IP dcache-onedata-01.desy.de.42855 > onedata.example.com.pcsync-https: Flags [S], seq 453727316, win 29200, options [mss 1460,sackOK,TS val 2063089930 ecr 0,nop,wscale 7], length 0
12:36:27.889975 IP onedata.example.com.pcsync-https > dcache-onedata-01.desy.de.42855: Flags [S.], seq 3527108424, ack 453727317, win 28960, options [mss 1460,sackOK,TS val 4086480795 ecr 2063089930,nop,wscale 7], length 0
12:36:27.890036 IP dcache-onedata-01.desy.de.42855 > onedata.example.com.pcsync-https: Flags [.], ack 1, win 229, options [nop,nop,TS val 2063089984 ecr 4086480795], length 0
12:36:27.903964 IP dcache-onedata-01.desy.de.42855 > onedata.example.com.pcsync-https: Flags [F.], seq 1, ack 1, win 229, options [nop,nop,TS val 2063089998 ecr 4086480795], length 0
12:36:27.958383 IP onedata.example.com.pcsync-https > dcache-onedata-01.desy.de.42855: Flags [.], ack 2, win 227, options [nop,nop,TS val 4086480865 ecr 2063089998], length 0
12:36:27.958408 IP onedata.example.com.pcsync-https > dcache-onedata-01.desy.de.42855: Flags [F.], seq 1, ack 2, win 227, options [nop,nop,TS val 4086480865 ecr 2063089998], length 0
12:36:27.958475 IP dcache-onedata-01.desy.de.42855 > onedata.example.com.pcsync-https: Flags [.], ack 2, win 229, options [nop,nop,TS val 2063090053 ecr 4086480865], length 0
```

So, it looks like some misconfiguration or bad hard-coded values are resulting in  our machine contacting `onedata.example.com`.

How do we proceed to fix this?

**Comment by groundnuty** [comment]
I've just successfully mounted the read only nfs share could you compare this working config with yours?

~~~
cluster:
  domainName: "a.cz"
  autoDeploy: true
  nodes:
    n1:
      hostname: "oneprovider"
  managers:
    mainNode: "n1"
    nodes:
      - "n1"
  workers:
    nodes:
      - "n1"
  databases:
    # Per node Couchbase cache size in MB for all buckets
    serverQuota: 4096
    # Per bucket Couchbase cache size in MB across the cluster
    bucketQuota: 1024
    nodes:
      - "n1"
  storages:
    onedatify:
      type: posix
      mountPoint: /onedatify/mnt/data-archive/
      readonly: true
    # Add initial storage resource (optional - can be added later)
    # In this example NFS mounted at /mnt/nfs on the host, which is
    # mounted to /volumes/storage directory inside Docker container
oneprovider:
  # Automatically register this Oneprovider in Onezone
  register: true
  name: "a"
  redirectionPoint: "https://oneprovider.a.cz"
  geoLatitude: 50.0848
  geoLongitude: 14.4112
onezone:
  # Assign custom name to the Onezone instance
  domainName: "a"
onepanel:
  # Create initially 1 administrator and 1 regular user
  users:
    "admin":
      password: "a"
      userRole: "admin"
~~~

**Comment by jstarek** [comment]
Hi all, thanks for the input. beta6 (and, subsequently, rc2) start up fine, so we can consider this issue solved. However, since chunked uploading still causes issues, I'll open another ticket (just to keep the issues properly separated).
